마지막 잎새가 떨어지기전 체크메이트... 여름이였다...

Choi-JJunho opened the pull request:
좋은데요?

woowahan-neo committed:
feat: remove web dependency

woowahan-neo committed:
feat: add example for functional programming api example

woowahan-neo committed:
chore: change package

Choi-JJunho committed:
docs: 커밋 메시지 템플릿 설정

Co-authored-by: jeomxon <jeomxon@gmail.com>

Choi-JJunho committed:
feat: 체스말과 체스판 구현

Co-authored-by: jeomxon <jeomxon@gmail.com>

Choi-JJunho committed:
feat: 체스 게임 시작과 종료 구현

Co-authored-by: jeomxon <jeomxon@gmail.com>

Choi-JJunho committed:
feat: View 추가

Co-authored-by: jeomxon <jeomxon@gmail.com>

Choi-JJunho committed:
feat: 체스판 View 출력 구현

Co-authored-by: jeomxon <jeomxon@gmail.com>

Choi-JJunho committed:
feat: 룩 이동 구현

Co-authored-by: jeomxon <jeomxon@gmail.com>

Choi-JJunho committed:
feat: 킹 이동 구현

Co-authored-by: jeomxon <jeomxon@gmail.com>

Choi-JJunho committed:
test: 빈 말에 대한 예외 검증

Co-authored-by: jeomxon <jeomxon@gmail.com>

Choi-JJunho committed:
feat: 비숍 이동 구현

Co-authored-by: jeomxon <jeomxon@gmail.com>

Choi-JJunho committed:
feat: 퀸 이동 구현

Co-authored-by: jeomxon <jeomxon@gmail.com>

Choi-JJunho committed:
feat: 나이트 이동 구현

Co-authored-by: jeomxon <jeomxon@gmail.com>

Choi-JJunho committed:
feat: 폰 이동 구현

Co-authored-by: jeomxon <jeomxon@gmail.com>

Choi-JJunho committed:
feat: 체스판 위에서의 룩 이동 구현

Co-authored-by: jeomxon <jeomxon@gmail.com>

Choi-JJunho committed:
feat: 체스판 위에서의 킹 이동 구현

Co-authored-by: jeomxon <jeomxon@gmail.com>

Choi-JJunho committed:
feat: 체스판 위에서의 비숍 이동 구현

Co-authored-by: jeomxon <jeomxon@gmail.com>

Choi-JJunho committed:
feat: 체스판 위에서의 나이트 이동 구현

Co-authored-by: jeomxon <jeomxon@gmail.com>

Choi-JJunho committed:
feat: 체스판 위에서의 폰,비숍 이동 구현

Co-authored-by: jeomxon <jeomxon@gmail.com>

Choi-JJunho committed:
feat: 체스게임 기능 구현

Co-authored-by: jeomxon <jeomxon@gmail.com>

Choi-JJunho committed:
mission: 스트림 미션 구현

Co-authored-by: jeomxon <jeomxon@gmail.com>

Choi-JJunho committed:
refactor: final 컨벤션 통일

Co-authored-by: jeomxon <jeomxon@gmail.com>

Choi-JJunho committed:
refactor: 테스트 Fixture 적용

Co-authored-by: jeomxon <jeomxon@gmail.com>

Choi-JJunho committed:
refactor: 코드 포맷팅

Co-authored-by: jeomxon <jeomxon@gmail.com>

Choi-JJunho committed:
refactor: 컨트롤러 로직 수정

Co-authored-by: jeomxon <jeomxon@gmail.com>

Choi-JJunho committed:
docs: 기능 목록 최신화

Co-authored-by: jeomxon <jeomxon@gmail.com>

be-student commented on build.gradle:
```diff
@@ -9,9 +9,6 @@ repositories {
 }
 
 dependencies {
-    implementation 'com.sparkjava:spark-core:2.9.3'
-    implementation 'com.sparkjava:spark-template-handlebars:2.7.1'
-    implementation 'ch.qos.logback:logback-classic:1.2.10'
```
이거 왜 변경되었죠?

Choi-JJunho replied:
글쎄요? 아마 인텔리제이 설정간 뭔가 일어난거 같은데 일단 필요는 없어보이네요,,,

drunkenhw commented on src/main/java/chess/board/ChessBoard.java:
```diff
@@ -0,0 +1,245 @@
+        }
+    }
+
+    // === createBoard ===
```
이 부분은 분리해도 좋지 않을까욥?

Choi-JJunho replied:
[ditto](https://github.com/woowacourse-code-review-study/java-chess/pull/4/files#r1140967723)

woo-chang commented on README.md:
```diff
@@ -5,3 +5,60 @@
+## 체스판
+
+- [x] 말의 위치를 관리한다.
+    - [x] 위치를 가진다.
+    - [x] 말을 가진다.
```
말의 위치를 관리한다라는 말로 아래의 의미가 담길 것 같은데 이 부분에 대해서는 어떻게 생각하시나요 🤔

위치를 가진다, 말을 가진다라는 말을 봤을 때 여러 개를 가지기보다는 `1`개만 관리한다는 느낌이 조금 들었습니다!

Choi-JJunho replied:
`말들의 위치를 가진다.` 로 문구를 수정했습니다.

woo-chang commented on README.md:
```diff
@@ -5,3 +5,60 @@
+
+## 말
+
+- [x] 색을 가진다
+    - [x] 흑 / 백
+
+- [x] 종류가 있다
+    - [x] 폰
+    - [x] 나이트
+    - [x] 비숍
+    - [x] 룩
+    - [x] 퀸
+    - [x] 킹
```
이런 자세한 설명은 너무 좋은데요 👍

woo-chang commented on src/main/java/chess/board/ChessBoard.java:
```diff
@@ -0,0 +1,245 @@
+        Piece fromPiece = piecePosition.get(from);
+        Piece toPiece = piecePosition.get(to);
+
+        if (fromPiece.isRook() && fromPiece.isMovable(from, to, toPiece)) {
+            validateFileAndRank(from, to);
+            move(from, to);
+        }
+
+        if (fromPiece.isBishop() && fromPiece.isMovable(from, to, toPiece)) {
+            validateDiagonal(from, to);
+            move(from, to);
+        }
+
+        if (fromPiece.isQueen() && fromPiece.isMovable(from, to, toPiece)) {
+            validateFileAndRank(from, to);
+            validateDiagonal(from, to);
+            move(from, to);
+        }
+
+        if (fromPiece.isPawn() && fromPiece.isMovable(from, to, toPiece)) {
+            validateRanks(from, to);
+            move(from, to);
+        }
+
+        if (fromPiece.isKing() && fromPiece.isMovable(from, to, toPiece)) {
+            move(from, to);
+        }
+
+        if (fromPiece.isKnight() && fromPiece.isMovable(from, to, toPiece)) {
+            move(from, to);
+        }
```
각 기물에 맞는 validate를 isMovable에 넣거나,

```java
if (fromPiece.isMovable(from, to, toPiece)) {
    ...
}
```

각 기물에 맞는 check로 override한다면 해당 메서드가

```java
if (fromPiece.isMovable(from, to, toPiece) && fromPiece.check(from, to)) {
    move(from, to);
}
```

위와 같이 간소화할 수 있고, 각 기물에 맞게 역할을 부여할 수 있을 것 같은데 어떻게 생각하시나요? 🤔

Choi-JJunho replied:
해당 부분에 대해서 각 기물에게 모든 검증을 부여하고자 한다면 가는 길에 말이 존재하는지에 대한 검증도 추가되어야하는데 이렇게 되면 ChessBoard와 Piece의 양방향 의존관계가 생기게 됩니다.ㅠㅠ

다즐이 말한것처럼 역할도 분리하고, 양방향 의존관계를 해결하는 방법을 열심히 찾아보고있습니다.
해결하고 공유하겠습니다.

woo-chang commented on src/main/java/chess/board/ChessBoard.java:
```diff
@@ -0,0 +1,245 @@
+
+    // === createBoard ===
+    private static void createPawn(final Map<Position, Piece> piecePosition, final Team team) {
+        if (team == Team.WHITE) {
+            for (final File file : File.values()) {
+                piecePosition.put(new Position(file, Rank.TWO), new Pawn(team));
+            }
+        }
+        if (team == Team.BLACK) {
+            for (final File file : File.values()) {
+                piecePosition.put(new Position(file, Rank.SEVEN), new Pawn(team));
+            }
+        }
```
매개변수로 `Rank`를 받으면 `Team` 확인할 필요없이 줄일 수 있을 것 같습니다!

```java
private static void createPawn(final Map<Position, Piece> piecePosition, final Rank rank, final Team team) {
    for (final File file : File.values()) {
        piecePosition.put(new Position(file, rank), new Pawn(team));}
    }
}
```

Choi-JJunho replied:
좋은데요?

woo-chang commented on src/main/java/chess/board/ChessBoard.java:
```diff
@@ -0,0 +1,245 @@
+        if (team == Team.BLACK) {
+            piecePosition.put(new Position(File.E, Rank.EIGHT), new King(team));
+        }
+    }
```
위의 메서드 전부 `ditto`

`ChessBoard` 코드를 확 줄일 수 있을 것 같은 느낌입니다!

woo-chang commented on src/main/java/chess/board/File.java:
```diff
@@ -0,0 +1,48 @@
+    E(5),
+    F(6),
+    G(7),
+    H(8),
```
파일은 영어 소문자로 표현해서 `character`를 사용해서 **a,b,c,d,e,f,g,h**로 두는건 어떻게 생각하시나요?

인덱스 계산도 동일하게 할 수 있습니다 :)

Choi-JJunho replied:
File을 표현할 때 이미 A,B,C,D... 를 나타내고 있기 때문에 해당 index에 대한부분은 숫자로 표현하는 것이 더 명확하다고 생각합니다.

woo-chang commented on src/main/java/chess/board/File.java:
```diff
@@ -0,0 +1,48 @@
+        this.index = index;
+    }
+
+    public static File of(final int index) {
```
`Nit`: 주노는 `from`보다 `of`를 선호하시나요?

be-student replied:
`FIY` : 인자가 하나면 from 이 좋아보여요

drunkenhw replied:
정팩메 컨벤션 읽어봅시다

Choi-JJunho replied:
[컨벤션](https://tecoble.techcourse.co.kr/post/2020-05-26-static-factory-method/) 읽고왔습니다.

죄송합니다.

woo-chang commented on src/main/java/chess/board/Position.java:
```diff
@@ -0,0 +1,36 @@
+
+import java.util.Objects;
+
+// TODO : 캐싱을 적용한다.
```
캐싱 적용하는거 좋은데요 기대됩니다 🔥

woo-chang commented on src/main/java/chess/board/Rank.java:
```diff
@@ -0,0 +1,48 @@
+import java.util.stream.IntStream;
+
+public enum Rank {
+    EIGHT(8),
+    SEVEN(7),
+    SIX(6),
+    FIVE(5),
+    FOUR(4),
+    THREE(3),
+    TWO(2),
+    ONE(1),
```
저도 거꾸로 뒤집는거 고민했었는데 사용하는 곳에서 뒤집어서 사용하는건 어떻게 생각하시나요?!

drunkenhw replied:
저도 뒤집긴했는데 순서가 바뀌어도 정상작동 되면 좋을 것 같네욥

Choi-JJunho replied:
뒤집은 이유가 DTO에서 values로 한번에 출력하고싶은 이유였어서
근거가 타당하지 않을 것 같아 정방향으로 정정했습니다.

죄송합니다.

woo-chang commented on src/main/java/chess/controller/ChessController.java:
```diff
@@ -0,0 +1,92 @@
+    }
+
+    private void processGame() {
+        final Command firstCommand = repeat(this::readCommand);
```
주노는 `repeat` 메서드명 어떻게 생각하시나요? 반복한다는게 강조되는 것 같아서 유효한 `Command`를 만드는 것에 대한 의미를 조금 더 담아보면 어떨까요?

Choi-JJunho replied:
어떠한 값을 검증 한 뒤 읽어온다는 의미로 `readValidateInput`이라는 이름으로 변경했습니다.

woo-chang commented on src/main/java/chess/controller/ChessController.java:
```diff
@@ -0,0 +1,92 @@
+
+    private Command readMoveCommand(final List<String> movePositions) {
+        return repeat(() -> {
+            final List<String> commands = inputView.readGameCommand();
+            final Command result = Command.from(commands.get(0));
+            if (result != Command.MOVE) {
+                return result;
+            }
+            movePositions.add(commands.get(1));
+            movePositions.add(commands.get(2));
+            return result;
```
외부에 존재하는 `movePositions`를 변경하는것에 대해서 어떻게 생각하시나요?

be-student replied:
안좋은데요?

Choi-JJunho replied:
인정합니다 컨트롤러를 너무 천대한 나머지 못난놈이 나왔네요

죄송합니다.

woo-chang commented on src/main/java/chess/controller/ChessController.java:
```diff
@@ -0,0 +1,92 @@
+    private Command readCommand() {
+        final List<String> commands = inputView.readGameCommand();
+        return Command.from(commands.get(0));
+    }
```
사용되지 않는 메서드인 것 같습니다 .. !

Choi-JJunho replied:
[여기](https://github.com/woowacourse-code-review-study/java-chess/pull/4/files#r1140944386)에서 사용됩니다.

안죄송합니다

woo-chang commented on src/main/java/chess/controller/ChessController.java:
```diff
@@ -0,0 +1,92 @@
+
+    private void renderChessBoard() {
+        final ChessBoard chessBoard = chessGame.getChessBoard();
+        final ChessBoardDto chessBoardDto = ChessBoardDto.toView(chessBoard);
```
dto 만드는 역할을 위임한 거 너무 좋은데요 👍

woo-chang commented on src/main/java/chess/controller/ChessController.java:
```diff
@@ -0,0 +1,92 @@
+        outputView.printChessBoard(chessBoardDto);
+    }
+
+    private <T> T repeat(final Supplier<T> function) {
+        Optional<T> input;
+        do {
+            input = repeatByEx(function);
+        } while (input.isEmpty());
+        return input.get();
+    }
+
+    private <T> Optional<T> repeatByEx(final Supplier<T> function) {
+        try {
+            return Optional.of(function.get());
+        } catch (IllegalArgumentException e) {
+            System.out.println(e.getMessage());
+            return Optional.empty();
+        }
+    }
```
`while`, `try` 쓰고 `depth 1`을 지키기 위한 방법이 있었네요 대박 👍

be-student replied:
와 이거 너무 좋은데요?

Choi-JJunho replied:
제 페어의 역작입니다.

제가 훔쳐왔습니다 죄송합니다

woo-chang commented on src/main/java/chess/dto/ChessBoardDto.java:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    private static List<PieceDto> pieceToView(final Map<Position, Piece> piecePosition) {
+        final List<PieceDto> pieceDtoList = new LinkedList<>();
```
`Nit`: `LinkedList`를 사용하신 이유가 궁금합니다!

Choi-JJunho replied:
지금을 flatMap을 이용하여 하나의 stream으로 묶어서 없애버렸습니다.

특별한 이유가 없었습니다.
죄송합니다

woo-chang commented on src/main/java/chess/dto/PieceDto.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    private String parseToPiece(final Piece piece) {
+        if (piece.isKing()) {
+            return "k";
+        }
+        if (piece.isQueen()) {
+            return "q";
+        }
+        if (piece.isKnight()) {
+            return "n";
+        }
+        if (piece.isPawn()) {
+            return "p";
+        }
+        if (piece.isRook()) {
+            return "r";
+        }
+        if (piece.isBishop()) {
+            return "b";
+        }
+        return ".";
```
저도 처음에 위와 같이 구현할까 고민하다가 `piece.getClass`를 활용해보았는데 이 방법은 어떻게 생각하시는지 궁금합니다!

해당 메서드를 사용하면

```java
Map<Class, String> map = new HashMap<>(Map.of(
    King.class, "k",
    Queen.class, "q",
    Knight.class, "n",
    Pawn.class, "p",
    Rook.class, "r",
    Bishop.class, "b"
));
```
`map`도 활용할 수 있었습니다!

be-student replied:
좋은데요?

Choi-JJunho replied:
확실히 getClass를 활용하면 작성에는 편리할 것 같네요!

이번에 저 또한 instanceof 혹은 .class를 이용할까 고민하다가, 객체간 메시지를 통해 상호작용을 한다는 부분에 초점을 두고 작업을 진행하여 isPawn, isKnight.. 등의 방식을 사용했습니다.

따라서 다음과 같은 방식으로 초기화를 고려해볼 수 있을 것 같네요

``` java
    private final Map<Predicate<Piece>, String> views = Map.of(
            Piece::isKing, "k",
            Piece::isQueen, "q",
            Piece::isKnight, "n",
            Piece::isPawn, "p",
            Piece::isRook, "r",
            Piece::isBishop, "b",
            Piece::isEmpty, "."
    );

...

    private String parseByTeam(final Piece piece2) {
        String result = views.entrySet().stream()
                .filter(piece -> piece.getKey().test(piece2))
                .findFirst()
                .map(Map.Entry::getValue)
                .orElseThrow(() -> new IllegalStateException("Piece 변환 과정 중 오류가 발생했습니다."));

        if (piece2.isBlack()) {
            return result.toUpperCase();
        }
        return result;
    }
```

좋은데요? 감사합니다

woo-chang commented on src/main/java/chess/game/ChessGame.java:
```diff
@@ -0,0 +1,37 @@
+public class ChessGame {
+
+    private final ChessBoard chessBoard;
+    private boolean processStatus;
```
`Command`를 가지지 않고 `status`로 변화시켜준 이유가 궁금합니다!

Choi-JJunho replied:
start 이후 start가 다시 입력되어도 종료되지 않게 끔 구성하려고 위와같이 구성했습니다.

woo-chang commented on src/main/java/chess/piece/Bishop.java:
```diff
@@ -0,0 +1,44 @@
+        final int fileInterval = File.calculateInterval(from.getFile(), to.getFile());
+        final int rankInterval = Rank.calculateInterval(from.getRank(), to.getRank());
+
+        if (fileInterval == rankInterval) {
+            return true;
+        }
+
+        throw new IllegalArgumentException("Bishop이 이동할 수 없는 경로입니다.");
```
`boolean`을 반환하기에

```java
return fileInterval == rankInterval
```

위를 반환하고, 외부에서 `false`인 경우 예외를 발생시키건 어떨까요?

Choi-JJunho replied:
수정사항이 엄청 커지긴 할거같은데 확실히 맞는말이네요..
구조 엎으면서 적용해보겠습니다.

woo-chang commented on src/main/java/chess/piece/King.java:
```diff
@@ -0,0 +1,43 @@
+        final boolean isFileIntervalOne = File.calculateInterval(from.getFile(), to.getFile()) <= 1;
+        final boolean isRankIntervalOne = Rank.calculateInterval(from.getRank(), to.getRank()) <= 1;
+
+        if (isFileIntervalOne && isRankIntervalOne) {
+            return true;
+        }
+        throw new IllegalArgumentException("King이 이동할 수 없는 경로입니다.");
```
각 기물마다 `ditto`입니다!

drunkenhw replied:
다즐과 ditto입니다

Choi-JJunho replied:
[뉴진스의 ditto](https://youtu.be/HtbnfBFIJls)입니다

woo-chang commented on src/main/java/chess/piece/Pawn.java:
```diff
@@ -0,0 +1,86 @@
+        throw new IllegalArgumentException("Pawn이 이동할 수 없는 경로입니다.");
+    }
+
+    private boolean isWhitePawnMovable(final Position from, final Position to, final Piece toPiece) {
+        final int fileInterval = File.calculateInterval(from.getFile(), to.getFile());
+
+        if (team == Team.WHITE) {
+
+            final int rankInterval = to.getRank().getIndex() - from.getRank().getIndex();
+
+            if (rankInterval == 1 && fileInterval == 1 && toPiece.isBlack()) {
+                return true;
+            }
+
+            if (from.getRank() == Rank.TWO) {
+                if (to.getRank().getIndex() - from.getRank().getIndex() <= 2 && to.getFile() == from.getFile()) {
+                    return true;
+                }
+                throw new IllegalArgumentException("Pawn이 이동할 수 없는 경로입니다.");
+            }
+
+            if (to.getRank().getIndex() - from.getRank().getIndex() == 1 && toPiece.isEmpty()) {
+                return true;
+            }
+        }
+        return false;
```
`Pawn`은 `Team`을 알고 있기에 이를 이용해서 아래 코드와 중복을 조금 줄여볼 수 없을까요? 🤔

drunkenhw replied:
아마 @be-student 눈우가 이 메소드를 밨다면 걷다가 빠져서 크게 다칠뻔 했을 것 같아욥~!

woo-chang commented on src/test/java/chess/fixture/FixturePosition.java:
```diff
@@ -0,0 +1,81 @@
+import chess.board.Position;
+import chess.board.Rank;
+
+public class FixturePosition {
```
테스트 픽스처 너무 좋습니다 👍

be-student replied:
좋은데요?

drunkenhw replied:
ㅋㅋ 노가다 열심히 하셨네욥 물론 이 부분은 저문이 했나욥?

Choi-JJunho replied:
합작입니다..

woo-chang reviewed (commented):
수고하셨습니다 :) 많이 배워갑니다 주노 고수 👾

be-student commented on src/main/java/Application.java:
```diff
@@ -0,0 +1,12 @@
+public class Application {
+
+    public static void main(final String[] args) {
+        final ChessController chessController = new ChessController(new InputView(), new OutputView(), new ChessGame());
```
바깥에서 game을 생성해서 넣어주네요
좋은 구조같기도 하네요

이 구조를 쓴 이유가 있나요?

Choi-JJunho replied:
향후 통합테스트를 작성할 때 유리하다고 생각하여 위와 같이 구성했습니다.

be-student commented on src/main/java/chess/view/OutputView.java:
```diff
@@ -0,0 +1,30 @@
+    public void printChessBoard(final ChessBoardDto chessBoardDto) {
+        final List<PieceDto> pieceDtos = chessBoardDto.getPieceDtos();
+
+        for (int i = 0; i < 64; i += 8) {
```
64나 8은 상수로 빼주세요

be-student commented on src/main/java/chess/piece/Queen.java:
```diff
@@ -0,0 +1,49 @@
+        }
+    }
+
+    @Override
+    public boolean isQueen() {
+        return true;
+    }
```
이렇게 각각 클래스마다 바깥에서 쓸 수 있도록 한다면, 다형성이 깨진게 아닐까요?

Choi-JJunho replied:
애매하네요 만나서이야기하시죠

be-student commented on src/main/java/chess/piece/Queen.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    private void validateStay(final Position from, final Position to) {
+        if (from.equals(to)) {
+            throw new IllegalArgumentException("제자리로는 움직일 수 없습니다.");
+        }
```
같은 에러가 계속 발생한다면 custom exception 을 만드는 것은 어떤가요?

Choi-JJunho replied:
좋은데요?

be-student commented on src/main/java/chess/piece/Knight.java:
```diff
@@ -0,0 +1,45 @@
+
+        if (isFileUnderTwo && isRankUnderTwo && fileInterval + rankInterval == 3) {
+            return true;
+        }
+        throw new IllegalArgumentException("Knight가 이동할 수 없는 경로입니다.");
+    }
```
이 부분도 
```java
class PieceMoveException extends RuntimeException{
public PieceMoveException(String pieceName){
super(pieceName+"가 이동할 수 없는 경로입니다");
}
}
```

Choi-JJunho replied:
ㅈㅇㄷㅇ?

be-student commented on src/main/java/chess/board/ChessBoard.java:
```diff
@@ -0,0 +1,245 @@
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+public class ChessBoard {
```
보드가 너무 많은 역할을 하고 있는 느낌이 들어요
board 의 역할을 조금 뺄 수 있을까요?
생성에 대한 역할을 다른 클래스로 빼는 팩토리도 가능할 것 같고
로직들도 뽑아볼 수 있을 것 같아요

drunkenhw replied:
tr;dr

Choi-JJunho replied:
생성 분리 좋은데요? 적용해보겠습니다.

+ @drunkenhw TL;DR입니다.

be-student reviewed (commented):
잘 봤어요 주노농

be-student commented on src/main/java/chess/dto/ChessBoardDto.java:
```diff
@@ -0,0 +1,41 @@
+
+    private final List<PieceDto> pieceDtos;
+
+    public static ChessBoardDto toView(final ChessBoard board) {
+        final List<PieceDto> pieces = pieceToView(board.getPiecePosition());
+        return new ChessBoardDto(pieces);
+    }
```
toView 라는 네이밍은 별로 안 좋은 것 같아요
이 네이밍 보다는 생성자에 두는 게 좋을지도요?

Choi-JJunho replied:
확실히 정팩메의 성질을 띄고 있는 것 같네요 `from`으로 정정했습니다

죄송합니다

drunkenhw commented on src/main/java/chess/board/ChessBoard.java:
```diff
@@ -0,0 +1,245 @@
+        Piece toPiece = piecePosition.get(to);
+
+        if (fromPiece.isRook() && fromPiece.isMovable(from, to, toPiece)) {
+            validateFileAndRank(from, to);
```
이런 말에 따라 달라지는 검증은 여기서 하는 이유가 있나욥?

Choi-JJunho replied:
경로에 대한 검증을 체스판에 넣으려다보니 지혜의 한계에 봉착했습니다.

좀더 쥐어 짜보겠습니다.

drunkenhw commented on src/main/java/chess/game/ChessGame.java:
```diff
@@ -0,0 +1,37 @@
+        chessBoard.movePiece(from, to);
+    }
+
+    public void receiveCommand(final Command command) {
+        if (command == Command.START) {
+            this.processStatus = true;
+        }
+        if (command == Command.END) {
+            this.processStatus = false;
+        }
```
시작 종료에 대한 상태를 여기에 둔 이유가 궁금해욥|~~~!

Choi-JJunho replied:
[ditto](https://github.com/woowacourse-code-review-study/java-chess/pull/4/files#r1140949015)

drunkenhw commented on src/main/java/chess/game/Command.java:
```diff
@@ -0,0 +1,23 @@
+package chess.game;
```
도메인이 view를 의존하고 있네욥~! 어케 생각하나욥

Choi-JJunho replied:
Command는 하위 도메인으로 깊이 넘어가는 부분이 아니라고 생각하여 지금으로서는 View로 분리해도 되려 과한 분리로 생각됩니다.

허리케인에 휩쓸려가도 안바꾸려고합니다

안죄송합니다

drunkenhw commented on src/main/java/chess/piece/Pawn.java:
```diff
@@ -0,0 +1,86 @@
+                }
+                throw new IllegalArgumentException("Pawn이 이동할 수 없는 경로입니다.");
+            }
+
```
ditto

drunkenhw commented on src/main/java/chess/piece/EmptyPiece.java:
```diff
@@ -0,0 +1,20 @@
+public class EmptyPiece extends Piece {
+
+    public EmptyPiece() {
+        super(null);
```
null? 언젠가 npe의 위험은 없을까요

Choi-JJunho replied:
확실히 나중에 메소드가 확장되고 team을 Empty에서 이용할 일이 생긴다면 NPE의 위험에 노출되어있겠네요...

EMPTY라는 상태를 추가해줬습니다.

죄송합니다

drunkenhw commented on src/test/java/chess/piece/BishopTest.java:
```diff
@@ -0,0 +1,41 @@
+        }
+
+        @Test
+        void 대각선이_아니면_예외() {
```
말이 짧다?

Choi-JJunho replied:
예외

drunkenhw commented on src/test/java/chess/piece/KingTest.java:
```diff
@@ -0,0 +1,40 @@
+            //given
+            King king = new King(Team.WHITE);
+
+            Position from = FixturePosition.A1;
+            Position to = FixturePosition.B2;
+
+            //when & then
+            assertThat(king.isMovable(from, to, PieceFixture.EMPTY_PIECE)).isTrue();
+        }
```
fixture변수를 따로 할당하지 않고 isMovable 파라미터에 A1,B2를 넣으면 가독성이 더 좋을 것 같다는 저만의 개인적인 생각입니다~

Choi-JJunho replied:
테스트 전반적인 통일성을 주려다보니 이렇게 되었네요~
바꾸지 않습니다

안죄송합니다

drunkenhw reviewed (commented):
안녕하세요 주노 박스타 입니다 
전반적으로 잘 읽었습니다 신림에서 시작해서 잠실새내까지 읽었네요 이제 저는 매머드 커피에서 아메리카노 사서 14층에 갈겁니다 그리고 제 코드 리팩토링 할거에요

Choi-JJunho committed:
docs: 요구사항 정의 수정

Choi-JJunho committed:
refactor: 말의 시작 Rank 분리 및 Rank에게 위임

Choi-JJunho committed:
refactor: 말 초기 생성 로직 다른 클래스로 분리

Choi-JJunho committed:
refactor: Rank 인덱스 순서로 재정렬

Choi-JJunho committed:
chore: 메소드명 변경

Choi-JJunho committed:
refactor: DTO의 if문 Map으로 변경

Choi-JJunho committed:
refactor: Team의 EMPTY 상태 추가

Choi-JJunho committed:
refactor: 상수 분리

Choi-JJunho committed:
refactor: Position 캐싱 적용

Choi-JJunho committed:
refactor: 말의 움직임에 대한 반환 방식 변경

- 커스텀 Exception인 PieceCannotMoveException 추가
- 기존 Piece의 구조 변경
  - isSomthing에서 Enum필드를 가지도록 변경
- 로직 변경에 따른 테스트 변경 및 추가

Choi-JJunho committed:
refactor: 테스트 Fixture 적용

Choi-JJunho committed:
chroe: 클래스명 변경

Choi-JJunho committed:
refactor: 말 객체 final 키워드 추가

Choi-JJunho committed:
refactor: 컨트롤러 로직 구현

- 패키지 구조 수정

Choi-JJunho committed:
refactor: 코드 포맷팅
